### ISSUE TITLE
Duplicate script calls

### ISSUE THREAD

**Issue by TehomCD** [evennia/evennia#1519]
#### Brief summary of issue / Description of requested feature:
Very rarely, I see the `at_repeat` for a script where there is only a single instance of it being called multiple times, usually with about a minute separation of it. 

#### Steps to reproduce the issue / Reasons for adding feature:

1. It's happened very rarely, and only after long periods of usage without the server being reloaded.

#### Error output / Expected result of feature
`at_repeat` is called multiple times, say twice within a minute or two of the first call.

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:
My guess is that a redundant `LoopingCall` is being created, but how or why is a complete mystery to me. It doesn't happen when the script is first created, because the script had run for days without interruption, and then we start to see the duplicate behavior, seemingly for no reason at all.

As a note, I did see this back in Evennia 0.6 as well, or at least I thought I did, with a different script. I noticed that a value which was supposed to be incremented at a script's `at_repeat` happened far, far more times than should have been possible. Since it was a timer, I just moved to datetime timedelta comparisons and thought no more of it. It was far more visible on the script I just saw because its `at_repeat` does a server broadcast, and I saw duplicate messages.

**Comment by Griatch** [comment]
@TehomCD I take it this is a global script. How long of a period is it using?

**Comment by TehomCD** [comment]
@Griatch Pretty long - 3600 seconds.

**Comment by TehomCD** [comment]
Noticed it happening after 2 days of uptime/between `@reload`s for a much shorter script - this one with a 300 second interval. It's pretty hard to notice unless the script does something that makes it obvious, like announcements during at_repeat.

**Comment by Griatch** [comment]
@TehomCD 

I'm running the following script on my local machine: 

```python
import time

class TestScript(Script):
    def at_script_creation(self):
        self.key = "test_repeat_script"
        self.desc = "test #1519"
        self.interval = 10
        self.persistent = True

    def at_start(self):
        self.ndb.started = time.time()
        self.ndb.repeats = 0

    def _get_expected(self):
        return abs(time.time() - self.ndb.started) // self.interval

    def at_server_reload(self):
        print("Reload: TestScript repeats/expected: %s/%s" % (self.ndb.repeats, self._get_expected()))

    def at_server_shutdown(self):
        print("Shutdown: TestScript repeats/expected: %s/%s" % (self.ndb.repeats, self._get_expected()))

    def at_repeat(self):
        if self.ndb.repeats != self._get_expected():
            print("TestScript repeats/expected: %s/%s" % (self.ndb.repeats, self._get_expected()))
        self.ndb.repeats += 1
```

This will reset on a reload which might be a potential problem. But I'll let it run for a while to see if I start seeing discrepancies between expected and actual repeats. You could maybe try to run the same to see if you see an effect. Suggestions for changes to make it more like your real-life case is welcome.

**Comment by TehomCD** [comment]
Since this has only happened after long periods of use, you might try making the cache flush at some point and seeing if that could be a cause also. I'm not sure if that's related, but it might be happening only after a cache flush.

In general, the duplicate at_repeat call was usually about a minute after the first one - not exactly, but around there. Might be worth including timestamps with the print statements so if it pops up you can see the exact time difference.

**Comment by TehomCD** [comment]
I'm trying something out to see if I can eliminate the duplicate `ExtendedLoopingCall`s. What I'm doing is overriding `at_idmapper_flush` for my Script typeclass and adding in something to stop the task of a flushed script, adding the current run time/callcount to a global variable dict, since I didn't want to do database writes during a flush. While this killed the tasks, the timer for validating scripts to restart them was up to an hour, so I added a callLater to a class method that calls the ScriptDB manager to validate them. However, without an active `self.ndb._task` the `start()` method wouldn't restart them while their `is_active` attribute was True, so I added a check in `start()` for _task being dead and then looking up in the dict to restart it with the appropriate values, analagous to unpausing. I'll see if this resolves the duplicate scripts, but I'm kinda hopeful.

**Comment by TehomCD** [comment]
I hadn't seen any duplicates in the days since I made the change locally in my typeclass, so I made it into the above PR to change `DefaultScript` that way.

Near as I can tell, the problem was that the ExtendedLoopingCall would persist after the script was flushed from memory - which makes sense if it's not explicitly stopped. So #1621 fixes that and handles edge cases I hadn't though of before Evennia's tests caught them, like Script objects that are being used as non-physical objects but have no LoopingCall associated with them.
